extensions/indexes/range/src/org/exist/indexing/range/RangeIndexConfigAttributeCondition.java

```diff
@@ -26,76 +26,380 @@
+                }
+
+        }
+
+        if (testStep != null && testValue != null) {
+            final QName qname = testStep.getTest().getName();
+            Comparable foundValue;
+            Comparable requiredValue;
+            boolean valueTypeMatches;
+
+            try {
+                if (this.numericComparison) {
+                    valueTypeMatches = testValue instanceof NumericValue;
+                    requiredValue = this.numericValue;
+                    foundValue = testValue.toJavaObject(Double.class);
+                }  else {
+                    valueTypeMatches = testValue instanceof StringValue;
+
+                    if (this.caseSensitive) {
+                        requiredValue = this.getLowercaseValue();
+                        foundValue = testValue.getStringValue().toLowerCase();
+                    } else {
+                        requiredValue = this.value;
+                        foundValue = testValue.getStringValue();
+                    }
+                }
+
+                if (qname.getNameType() == ElementValue.ATTRIBUTE
+                        && operator.equals(this.operator)
+                        && qname.equals(this.attribute)
+                        && valueTypeMatches
+                        && foundValue.equals(requiredValue)) {
 
-                if (operator.equals(this.operator) && attribute.equals(this.attribute) && value.equals(value)) {
                     return true;
                 }
+
+
+            } catch (XPathException e) {
+                RangeIndex.LOG.error("Value conversion error when testing predicate for condition, value: " + testValue.toString());
+                RangeIndex.LOG.error(e);
             }
+
         }
 
+        return false;
+    }
 
+    private Expression unwrapSubExpression(Expression expr) {
 
-        return false;
+        if (expr instanceof Atomize) {
+            expr = ((Atomize) expr).getExpression();
+        }
+
+        if (expr instanceof DynamicCardinalityCheck) {
+            if (expr.getSubExpressionCount() == 1) {
+                expr = expr.getSubExpression(0);
+            }
+        }
+
+        if (expr instanceof PathExpr) {
+            if (expr.getSubExpressionCount() == 1) {
+                expr = expr.getSubExpression(0);
+            }
+        }
+
+        return expr;
+    }
+
+    private LocationStep findLocationStep(Expression expr) {
+
+        if (expr instanceof LocationStep) {
+            return (LocationStep) expr;
+        }
+
+        return null;
+    }
+
+    private AtomicValue findAtomicValue(Expression expr) {
+
+        if (expr instanceof AtomicValue) {
+            return (AtomicValue) expr;
+        }
+        else if (expr instanceof LiteralValue) {
+            return ((LiteralValue) expr).getValue();
+
+        } else if (expr instanceof VariableReference || expr instanceof Function) {
+            try {
+                final Sequence result = expr.eval(null);
```

Comment:
You don't need to worry about the security, the same user as executed the query will be used.